Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp

```diff
@@ -32,70 +32,113 @@
 #include "CSSPropertyParserConsumer+Color.h"
 #include "CSSPropertyParserConsumer+Ident.h"
 #include "CSSPropertyParserConsumer+Length.h"
+#include "CSSPropertyParserConsumer+LengthDefinitions.h"
 #include "CSSPropertyParserConsumer+List.h"
+#include "CSSPropertyParserConsumer+MetaConsumer.h"
 #include "CSSPropertyParserConsumer+String.h"
-#include "CSSShadowValue.h"
+#include "CSSTextShadowPropertyValue.h"
 #include "CSSValueList.h"
 #include "CSSValuePair.h"
 
 namespace WebCore {
 namespace CSSPropertyParserHelpers {
 
-static RefPtr<CSSValue> consumeSingleTextShadow(CSSParserTokenRange& range, const CSSParserContext& context)
+static std::optional<CSS::TextShadow> consumeSingleUnresolvedTextShadow(CSSParserTokenRange& range, const CSSParserContext& context)
 {
     // <single-text-shadow> = [ <color>? && <length>{2,3} ]
     // https://drafts.csswg.org/css-text-decor-3/#propdef-text-shadow
 
-    RefPtr<CSSValue> color;
-    RefPtr<CSSPrimitiveValue> horizontalOffset;
-    RefPtr<CSSPrimitiveValue> verticalOffset;
-    RefPtr<CSSPrimitiveValue> blurRadius;
+    // FIXME: CSS Text Decoration 4 has updated text-shadow to use the complete box-shadow grammar:
+    // <shadow> = <color>? && [<length>{2} <length [0,∞]>? <length>?] && inset?
+    // https://drafts.csswg.org/css-text-decor-4/#propdef-text-shadow
+
+    const auto lengthOptions = CSSPropertyParserOptions {
+        .parserMode = context.mode,
+        .unitlessZero = UnitlessZeroQuirk::Allow
+    };
+
+    std::optional<CSS::Color> color;
+    std::optional<CSS::Length<>> x;
+    std::optional<CSS::Length<>> y;
+    std::optional<CSS::Length<CSS::Nonnegative>> blur;
+
+    auto consumeOptionalColor = [&] -> bool {
+        if (color)
+            return false;
+        auto maybeColor = consumeUnresolvedColor(range, context);
+        if (!maybeColor)
+            return false;
+        color = CSS::Color(WTFMove(*maybeColor));
+        return !!color;
+    };
+
+    auto consumeLengths = [&] -> bool {
+        if (x)
+            return false;
+        x = MetaConsumer<CSS::Length<>>::consume(range, context, { }, lengthOptions);
+        if (!x)
+            return false;
+        y = MetaConsumer<CSS::Length<>>::consume(range, context, { }, lengthOptions);
+        if (!y)
+            return false;
```

Comment:
It's OK here that we consumed `x` as a side effect even though we returned `false`? I find these functions easier to understand when they either succeed or do nothing but I guess there are many variations.

Reply:
In this case, the only caller is doing a save/restore of the range based on the whole function, so it is ok, but I agree that making each function self sufficient in this regard is much better and makes refactoring more straightforward. I will fix this.